src/Lykke.Job.CandlesHistoryWriter.Services/HistoryMigration/TradesMigrationManager.cs

```diff
@@ -0,0 +1,139 @@
+                        foreach (var interval in Constants.StoredIntervals)
+                        {
+                            // It's important for Constants.StoredIntervals to be ordered by time period increase,
+                            // because we will calculate candles for each period based on previous period candles.
+                            var currentCandles = interval == CandleTimeInterval.Sec
+                                ? new TradesCandleBatch(migrationItem.AssetId, interval, tradesBatch)
+                                : new TradesCandleBatch(migrationItem.AssetId, interval, smallerCandles);
+
+                            ExtendStoredCandles(ref currentCandles);
+
+                            Health.AssetReportItems[migrationItem.AssetId].SummarySavedCandles += currentCandles.CandlesCount;
+
+                            smallerCandles = currentCandles;
+
+                            // Preffered to perform inserting in parralel style.
+                            batchInsertTasks.Add(
+                                _candlesHistoryRepository.InsertOrMergeAsync(
+                                    currentCandles.Candles.Values,
+                                    currentCandles.AssetId,
+                                    TradesCandleBatch.PriceType,
+                                    currentCandles.TimeInterval));
+                        }
+
+                        Health.AssetReportItems[migrationItem.AssetId].SummaryFetchedTrades += batchCount;
+
+                        Task.WaitAll(batchInsertTasks.ToArray());
+
+                        await _log.WriteInfoAsync(nameof(TradesMigrationManager), nameof(Migrate),
+                            $"Batch of {batchCount} records for {migrationItem.AssetId} processed.");
+                    }
+
+                    Health.State = TradesMigrationState.Finished;
+
+                    await _log.WriteInfoAsync(nameof(TradesMigrationManager), nameof(Migrate),
+                        $"Migration for {migrationItem.AssetId} finished. Total records migrated: {Health.AssetReportItems[migrationItem.AssetId].SummaryFetchedTrades}, total candles stored: {Health.AssetReportItems[migrationItem.AssetId].SummarySavedCandles}.");
+                }
+            }
+        }
+
+        private void ExtendStoredCandles(ref TradesCandleBatch current)
+        {
+            var storedCandles =
+                _candlesHistoryRepository.GetCandlesAsync(current.AssetId, current.TimeInterval, TradesCandleBatch.PriceType, current.MinTimeStamp,
+                    current.MaxTimeStamp.AddSeconds((int)current.TimeInterval)).GetAwaiter().GetResult().ToList();
+            if (!storedCandles.Any())
+                return;
+
+            for (int i = 0; i < current.CandlesCount; i++)
+            {
+                var timestamp = current.Candles.Values.ElementAt(i).Timestamp;
+                var stored = storedCandles.FirstOrDefault(s => s.Timestamp == timestamp);
```

Comment:
If you make ToDictionary(x => x.Timestamp, x => x) with result of the GetCandlesAsync(), then you could use storedCandles.TryGetValue(timestamp, var out stored). This is much better than linear search O(N). And you don't need to make storedCandles.Remove(stored) then, which is O(N) too.